### PR TITLE
feat(caip): onramper support

### PR DESCRIPTION
**Pull request by lmyslinski**
### Description

Full lib-side `OnRamper` support. Because they support multiple payment gateways that use different id's we need to use a more complex mapping to make sure all gateways are included in our request

### Issue

- N/A - needed for https://github.com/shapeshift/web/pull/2583 web PR

### CODE REVIEW

File: packages/caip/src/adapters/onramper/index.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  'eip155:1/erc20:0x0f5d2fb29fb7d3cfee444a200298f468908cc942': 'MANA_ERC20',
+  'eip155:1/erc20:0xd26114cd6ee289accf82350c8d8487fedb8a0c07': 'OMG',
+  'eip155:1/erc20:0x57ab1ec28d129707052df4df418d58a2d46d5f51': 'SUSDC',
+  'eip155:1/erc20:0x6b3595068778dd592e39a122f4f5a5cf09c90fe2': 'SUSHI_ERC20',
+  'eip155:1/erc20:0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f': 'SNX',
+  'eip155:1/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f984': 'UNI-ERC20',
+  'eip155:1/erc20:0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48': 'USDC',
+  'eip155:1/erc20:0xdac17f958d2ee523a2206206994597c13d831ec7': 'USDT',
+  'eip155:1/erc20:0x2260fac5e5542a773aa44fbcfedf7c193bc2c599': 'WBTC',
+} as Record<AssetId, string>
```
Comment (0xdef1cafe) [inline packages/caip/src/adapters/onramper/index.ts:34]:
```suggestion
const AssetIdToTransakIdMap: Record<AssetId, string> = {
  [btcAssetId]: 'BTC',
  [cosmosAssetId]: 'ATOM',
  [ethAssetId]: 'ETH',
  [avalancheAssetId]: 'AVAX',
  'eip155:1/erc20:0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9': 'AAVE_ECR20',
  'eip155:1/erc20:0xbb0e17ef65f82ab018d8edd776e8dd940327b28b': 'AXS',
  'eip155:1/erc20:0x4d224452801aced8b2f0aebe155379bb5d594381': 'APE',
  'eip155:1/erc20:0x0d8775f648430679a709e98d2b0cb6250d2887ef': 'BAT_ERC20',
  'eip155:1/erc20:0x4fabb145d64652a948d72533023f6e7a623c7c53': 'BUSD',
  'eip155:1/erc20:0x3506424f91fd33084466f402d5d97f05f8e3b4af': 'CHZ',
  'eip155:1/erc20:0xc00e94cb662c3520282e6f5717214004a7f26888': 'COMP',
  'eip155:1/erc20:0x6b175474e89094c44da98b954eedeac495271d0f': 'DAI_ERC20',
  'eip155:1/erc20:0xf629cbd94d3791c9250152bd8dfbdf380e2a3b9c': 'ENJ',
  'eip155:1/erc20:0x4e15361fd6b4bb609fa63c81a2be19d873717870': 'FTM',
  'eip155:1/erc20:0xf57e7e7c23978c3caec3c3548e3d615c346e79ff': 'IMX',
  'eip155:1/erc20:0x514910771af9ca656af840dff83e8264ecf986ca': 'LINK',
  'eip155:1/erc20:0x9f8f72aa9304c8b593d555f12ef6589cc3a579a2': 'MKR',
  'eip155:1/erc20:0x0f5d2fb29fb7d3cfee444a200298f468908cc942': 'MANA_ERC20',
  'eip155:1/erc20:0xd26114cd6ee289accf82350c8d8487fedb8a0c07': 'OMG',
  'eip155:1/erc20:0x57ab1ec28d129707052df4df418d58a2d46d5f51': 'SUSDC',
  'eip155:1/erc20:0x6b3595068778dd592e39a122f4f5a5cf09c90fe2': 'SUSHI_ERC20',
  'eip155:1/erc20:0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f': 'SNX',
  'eip155:1/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f984': 'UNI-ERC20',
  'eip155:1/erc20:0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48': 'USDC',
  'eip155:1/erc20:0xdac17f958d2ee523a2206206994597c13d831ec7': 'USDT',
  'eip155:1/erc20:0x2260fac5e5542a773aa44fbcfedf7c193bc2c599': 'WBTC',
}
```

always assign over cast

File: packages/caip/src/adapters/onramper/index.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+const onRamperTickerToAssetIdMap = invert(AssetIdToTransakIdMap)
+
+export const getOnRamperSupportedAssets = () =>
+  entries(AssetIdToTransakIdMap).map(([assetId, ticker]) => ({
+    assetId,
+    ticker,
+  }))
+
+export const onRamperTickerToAssetId = (id: string): string | undefined =>
```
Comment (0xdef1cafe) [inline packages/caip/src/adapters/onramper/index.ts:44]:
```suggestion
export const onRamperTickerToAssetId = (id: string): AssetId | undefined =>
```

File: packages/caip/src/adapters/onramper/index.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  onRamperTickerToAssetIdMap[id]
+
+export const assetIdToOnRamperTicker = (assetId: string): string | undefined =>
```
Comment (0xdef1cafe) [inline packages/caip/src/adapters/onramper/index.ts:47]:
```suggestion
export const assetIdToOnRamperTicker = (assetId: AssetId): string | undefined =>
```

**Review by gomesalexandre — COMMENTED** [review]
Tested linked with the matching web PR both with native wallet and MetaMask:

- ETH/ERC20s/AVAX/.e
- Cosmos and UTXOs (only Bitcoin as it's the only supported UTXO in fiat ramps currently)
- Buy/Sell tabs

Noticed some issues WRT the selected sell asset / minimum sell amount, but those are web concerns.

LGTM, will stamp this one after the typo is fixed

File: packages/caip/src/adapters/onramper/index.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  'eip155:1/erc20:0x0f5d2fb29fb7d3cfee444a200298f468908cc942': ['MANA', 'MANA_ERC20'],
+  'eip155:1/erc20:0xd26114cd6ee289accf82350c8d8487fedb8a0c07': ['OMG'],
+  'eip155:1/erc20:0x57ab1ec28d129707052df4df418d58a2d46d5f51': ['SUSDC'],
+  'eip155:1/erc20:0x6b3595068778dd592e39a122f4f5a5cf09c90fe2': ['SUSHI', 'SUSHI_ERC20'],
+  'eip155:1/erc20:0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f': ['SNX'],
+  'eip155:1/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f984': ['UNI', 'UNI-ERC20'],
+  'eip155:1/erc20:0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48': ['USDC'],
+  'eip155:1/erc20:0xdac17f958d2ee523a2206206994597c13d831ec7': ['USDT'],
+  'eip155:1/erc20:0x2260fac5e5542a773aa44fbcfedf7c193bc2c599': ['WBTC'],
+}
```
Comment (gomesalexandre) [inline packages/caip/src/adapters/onramper/index.ts:35]: This will fail CI
```suggestion
}
```

**Comment by lmyslinski** [comment]
@gomesalexandre Thanks for the comments, all resolved

File: packages/caip/src/adapters/onramper/index.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+// explodes and inverts the assetId => tokenId[] map by creating a 1to1 token => assetId mapping
+const invertMap = () => {
+  const invertedMap: Record<OnRamperTokenId, AssetId> = {}
+  entries(AssetItToOnRamperIdMap).flatMap(([assetId, idList]) =>
```
Comment (gomesalexandre) [inline packages/caip/src/adapters/onramper/index.ts:40]: Still present here
```suggestion
  entries(AssetIdToOnRamperIdMap).flatMap(([assetId, idList]) =>
```

File: packages/caip/src/adapters/onramper/index.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      invertedMap[id] = assetId
+    }),
+  )
+  return invertedMap
+}
+
+const OnRamperIdToAssetIdMap = invertMap()
+
+export const getOnRamperSupportedAssets = () => {
+  entries(AssetItToOnRamperIdMap).map(([assetId, tokenId]) => ({
```
Comment (gomesalexandre) [inline packages/caip/src/adapters/onramper/index.ts:51]: Still present here
```suggestion
  entries(AssetIdToOnRamperIdMap).map(([assetId, tokenId]) => ({
```

File: packages/caip/src/adapters/onramper/index.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    assetId,
+    token: tokenId,
+  }))
+}
+
+export const onRamperTokenIdToAssetId = (tokenId: OnRamperTokenId): AssetId | undefined => OnRamperIdToAssetIdMap[tokenId]
+
+export const assetIdToOnRamperTokenList = (assetId: AssetId): OnRamperTokenId[] | undefined => AssetItToOnRamperIdMap[assetId]
```
Comment (gomesalexandre) [inline packages/caip/src/adapters/onramper/index.ts:59]: Still present here

```suggestion
export const assetIdToOnRamperTokenList = (assetId: AssetId): OnRamperTokenId[] | undefined =>
  AssetIdToOnRamperIdMap[assetId]
```

File: packages/caip/src/adapters/onramper/index.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+const OnRamperIdToAssetIdMap = invertMap()
+
+export const getOnRamperSupportedAssets = () => {
+  entries(AssetItToOnRamperIdMap).map(([assetId, tokenId]) => ({
+    assetId,
+    token: tokenId,
+  }))
+}
+
+export const onRamperTokenIdToAssetId = (tokenId: OnRamperTokenId): AssetId | undefined => OnRamperIdToAssetIdMap[tokenId]
```
Comment (gomesalexandre) [inline packages/caip/src/adapters/onramper/index.ts:57]: Lint

```suggestion
export const onRamperTokenIdToAssetId = (tokenId: OnRamperTokenId): AssetId | undefined =>
  OnRamperIdToAssetIdMap[tokenId]
```

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
This currently fails CI as typo wasn't fixed consistently.
After fixing all occurences, lint will also fail, please run `yarn lint` and ideally set-up eslint in your editor of choice to auto-spot/fix all linter violations

**Review by gomesalexandre — APPROVED** [review]
Tested again while linked in web, let's get this one in 🍬

**Pull request merged**